src/fp/ops.rs

```diff
@@ -0,0 +1,417 @@
+        // +  r1,r0
+        //    =====
+        // cc, z, 0
+        let (_zero, carry) = z0.overflowing_add(&r0);
+        let (cc, z) =
+            hi_lo(z1.as_() + r1.as_() + <Self::DoubleWord as Word>::Carry::from(carry).as_());
+
+        // Final subtraction
+        // If z >= p, then z = z - p
+
+        //    cc, z
+        // -   0, p
+        // ========
+        // b1,s1,s0
+        let (s0, b0) = z.overflowing_sub(&Self::PRIME);
+        let (_s1, b1) = cc.overflowing_sub(&W::Carry::from(b0).as_());
+        // if b1 == 1: return z
+        // else:       return s0
+        let mask = W::ZERO.wrapping_sub(&W::Carry::from(b1).as_());
+        (z & mask) | (s0 & !mask)
+    }
+}
+
+/// `FieldMulOpsTwoWords` implements prime field multiplication.
+///
+/// The implementation assumes that the modulus `p` fits in one word of
+/// 'W' bits, but the product of two integers does not fit. Thus,
+/// multiplication is processed splitting integers in two words.
+pub(crate) trait FieldMulOpsSplitWord<W: Word>: FieldParameters<W> {
+    /// Multiplication of field elements in the Montgomery domain.
+    /// This uses the Montgomery's [REDC algorithm][montgomery].
+    /// The result will be in [0, p).
+    ///
+    /// [montgomery]: https://www.ams.org/journals/mcom/1985-44-170/S0025-5718-1985-0777282-X/S0025-5718-1985-0777282-X.pdf
+    fn mul(x: W, y: W) -> W {
+        let high = |v: W| v >> (W::BITS / 2);
+        let low = |v: W| v & ((W::ONE << (W::BITS / 2)) - W::ONE);
+
+        let (x1, x0) = (high(x), low(x));
+        let (y1, y0) = (high(y), low(y));
+
+        // Integer multiplication
+        // z = x * y
+
+        //       x1,x0
+        // *     y1,y0
+        // ===========
+        // z3,z2,z1,z0
+        let mut result = x0 * y0;
+        let mut carry = high(result);
+        let z0 = low(result);
+        result = x0 * y1;
+        let mut hi = high(result);
+        let mut lo = low(result);
+        result = lo + carry;
+        let mut z1 = low(result);
+        let mut cc = high(result);
+        result = hi + cc;
+        let mut z2 = low(result);
+
+        result = x1 * y0;
+        hi = high(result);
+        lo = low(result);
+        result = z1 + lo;
+        z1 = low(result);
+        cc = high(result);
+        result = hi + cc;
+        carry = low(result);
+
+        result = x1 * y1;
+        hi = high(result);
+        lo = low(result);
+        result = lo + carry;
+        lo = low(result);
+        cc = high(result);
+        result = hi + cc;
+        hi = low(result);
+        result = z2 + lo;
+        z2 = low(result);
+        cc = high(result);
+        result = hi + cc;
+        let mut z3 = low(result);
+
+        // Montgomery Reduction
+        // z = z + p * mu*(z mod 2^64), where mu = (-p)^(-1) mod 2^64.
+
+        // z3,z2,z1,z0
+        // +     p1,p0
+        // *         w = mu*z0
+        // ===========
+        // z3,z2,z1, 0
+        let mut w = low(Self::MU.wrapping_mul(&z0));
```

Comment:
Previously, `FieldParameters::mu` was a `u64` field for 128-bit primes. Now, the `FieldParameters::MU` associated constant has type `W`, meaning it's a `u128` for 128-bit primes. The previous code had a `as u64` cast on the right multiplicand. For one thing, this means the wrapping multiplication operations will wrap at different powers of two. Thus, I'm worried there could be correctness issues downstream if `w` is not in the range it's supposed to be. Secondly, from a performance perspective, this would now be a u128 * u128 multiplication where we expect the top half of both multiplicands to be zero, so the compiler would emit a built-in with three extra u64 * u64 multiplications we're not truly using.

To solve this, we may be able to pass both inputs through `low()`, and then pass the result through `low()` as well to wrap at the right place. It's possible the compiler will see all the masking operations and emit the wrapping u64 * u64 multiplication we want. Alternately, the more involved and more complete solution would be to add another type parameter or an associated for unsigned integers with LOG2_BASE bits, as opposed to `W` which has LOG2_RADIX bits. Then we could make MU smaller, and add the truncating cast back here. (one implementation with 'words' and 'double words', another with 'half words' and 'words')

Reply:
Good catch, I've added an auxiliary type and add a compile-time check that tests the size of MU.